### ISSUE TITLE
Support track-sc in http-after-response

### ISSUE THREAD

**Issue by soapyp** [haproxy/haproxy#2628]
### Your Feature Request

HAProxy should allow the use of track-sc* in http-after-response directives.

### What are you trying to do?

I am trying to track HTTP response codes associated with a client API key in a stick table, including from responses issued by HAProxy such as in `http-request deny` directives and internal errors. I can't set up the counter in `http-request` as part of the key is the status code, which is not available there. I also can't set up the counter in `http-response` as denied requests skip these rules. I can correctly get the status and generate my stick table key in `http-after-response`, but although I can use GPCs here I can't set up the counter as the `track-sc*` keywords are not supported.

I read a Github issue (https://github.com/haproxy/haproxy/issues/1980#issuecomment-1370045200) where a contributor added `sc-inc-gpc` among other keywords to `http-after-response`, but neglected the counter setup due to performance concerns at the end of the request lifecycle, and encouraging it's use for the 'wrong reason'. If there's another way you can think of to achieve this goal I would be more than happy to hear it and close the issue.

I briefly looked into how large of a change this would be (being totally unfamiliar with the codebase so please forgive me if it's more complex) and I _think_ it's an one line addition in `http_act.c`. I am happy to make the change, update any tests/docs and submit a PR if this is something you would consider merging.
Please consider my request as this felt like something which should be easily achievable, but have spent a whole day editing configuration in an attempt to dance around this seemingly arbitrary restriction.

Many Thanks

### Output of `haproxy -vv`

```plain
HAProxy version 2.9.9-ad75c48 2024/06/14 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.9.9.html
Running on: Linux 5.15.98-fly #gac12cf665d SMP Thu Jun 20 16:59:12 UTC 2024 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1 USE_PCRE2=1 USE_PCRE2_JIT=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC -QUIC_OPENSSL_COMPAT +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=2).
Built with OpenSSL version : OpenSSL 3.0.13 30 Jan 2024
Running on OpenSSL version : OpenSSL 3.0.13 30 Jan 2024
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.4.4
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.42 2022-12-11
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 12.2.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

**Comment by wtarreau** [comment]
Hello,

I'm a bit puzzled by reading that "a part of the key is the status code", because 1) this means that haproxy itself can alter the key, and 2) you will never be able to match that key at any other moment than in this
after-response ruleset, then I can't imagine any action that could result from this tracking.

I sense that instead you're trying to etablish some correlation between the client API key (present in the request), the status (present in the response), and store information and/or statistics based on this. Am I wrong?

If that's the case, I tend to think that your tracking key should be the client API key, and that you would just have to use different GPC for different status codes. This way for a single key (a "line" in the table) you'd have all your counters at once. We could for example imagine that you increment gpc1 for 1xx status codes, gpc2 for 2xx, gpc3 for 3xx, gpc4 for 4xx, gpc5 for 5xx. Does it sound like it would match your needs?

You may notice I'm reluctant to allowing track-sc in http-after-response, but the reason is that track-sc involves numerous implicit counters when done late, that we have to try to figure and cover for. In your case, in after-response rules for example, errors were already processed so we no longer have the distinction between denies and failures for example, so anyone using that would have garbage in such counters. So that's quite hackish and I can't say what will work fine, what will sometimes work and what will never work.

**Comment by soapyp** [comment]
Hi @wtarreau,

Yes, the goal was to correlate client keys with ultimate response statuses and store them in stick tables. I needed this purely for monitoring purposes, so wasn't worried about not being able to perform actions in HAProxy based on the stored data. I took another run at using the GPCs after reading your message. Previously they didn't seem to work and I attributed that to having to set up the sticky counter in a different part of the request lifecycle. I've since found that the issue was in my prometheus exporter and after making some changes there it works like a charm!

I appreciate the reasoning for your reluctance to add track-sc to http-after-response. I underestimated the complexity involved, so apologies if I wasted some of your valuable time. Hope others can use the advice here if they find themselves in a similar situation.

Closing this issue for now as my problem was solved without the need for a new feature.

Many thanks for your help!
